sematic/ui/src/components/ScrollingLogView.tsx

```diff
@@ -2,180 +2,40 @@ import { useEffect, useState, useCallback, useMemo } from "react";
-  // are currently set, load new logs with the current settings.
-  useEffect(() => {
-    if (
-      lineState.source !== logSource ||
-      lineState.filterString !== filterString
-    ) {
-      next();
-    }
-  });
-
-  const noMoreLinesIndicator = (
-    <Alert severity="info" sx={{ mt: 3 }}>
-      {currentNoLinesReason}
-    </Alert>
-  );
-  const scrollerId = "scrolling-logs-" + lineState.source;
-
-  // repeatedly query the server until we have loaded as many log lines as
-  // it will give us.
-  const accumulateUntilEnd = useCallback(() => {
-    if (lineState.lines.length === 0) {
-      // run has no logs. No need to try to scroll to the end.
-      return;
-    }
-    setFastForwarding(true);
-    var accumulatedLines: string[] = [];
-    const accumulate = function (
-      source: string,
-      usedFilter: string,
-      lines: string[],
-      cursor: string | null,
-      noLinesReason: string | null
-    ) {
-      accumulatedLines.push(...lines);
-      if (
-        (cursor === null || lines.length === 0) &&
-        accumulatedLines.length > 0
-      ) {
-        setLoadingMessage("Rendering...");
-
-        setFastForwarding(false);
-        handleLogLines(
-          source,
-          usedFilter,
-          accumulatedLines,
-          cursor,
-          noLinesReason
-        );
-
-        // start the scroll-to-bottom asynchronously so the lines have time
-        // to actually render before it scrolls.
-        setTimeout(() => {
-          const scroller = document.getElementById(scrollerId);
-          scroller?.scrollTo(0, scroller.scrollHeight);
-        }, 0);
-      } else {
-        setLoadingMessage("Loaded " + accumulatedLines.length + " lines...");
-        getLines(source, cursor, usedFilter, accumulate);
-      }
-    };
-    accumulate(logSource, filterString, [], null, null);
-  }, [getLines, handleLogLines, logSource, scrollerId]);
-
-  // scroll to the bottom when fast forwarding/jumping to end is done
-  useMemo(() => {
-    if (!fastForwarding) {
-      const scroller = document.getElementById(scrollerId);
-      scroller?.scrollTo(0, scroller.scrollHeight);
+  const scrollerId = useMemo(() => `scrolling-logs-${logSource}`, [logSource]) ;
+
+  // Single pull logic
+  const { lines, isLoading, error, hasMore, 
+    noMoreLinesReason, getNext, hasPulledData } = useLogStream(logSource, filterString);
+  
+  // Accumulator (logs draining) logic
+  const { accumulateLogsUntilEnd, isLoading: isAccumulatorLoading,
+    isAccumulating, accumulatedLines } = useAccumulateLogsUntilEnd(hasMore, getNext);
+  
+  // report to parent in case of errors.
+  useEffect(()=> {
+    if (!!error) {
+      onError(error);
     }
-  }, [fastForwarding, scrollerId]);
+  }, [onError, error])
 
   const onFilterStringChange = useCallback(
     (evt: any) => {
       setFilterString(evt.target.value);
     },
-    [filterString, next]
+    [setFilterString]
   );
 
   const onScroll = useCallback(
```

Comment:
It is unclear to me whether we still need this callback. 

`pullDownToRefresh` and `pullDownToRefreshThreshold` from [react-infinite-scroll-component](https://github.com/ankeetmaini/react-infinite-scroll-component) seems to fulfill this purpose. Was there any issue encountered using that? 

In reality, I never saw this `refresh` behavior get triggered, the infinite scrolling behavior always occurs instead. UX-wise, I haven't found any noticeable issues.

Also, I am not confident that I fully grasp this comment paragraph. So maybe there is some misunderstanding.

Maybe the original author, @augray, could share some insight into this?

Reply:
This was intended to enable a behavior where you could “yank” the logs by scrolling to hit the end *while the logs were being actively generated* to load the latest logs. I think there was a PR that broke that behavior, but it would be good to get it working again. Developing it is a bit tough since you have to do it with a run that is continuously producing logs, but I believe @tscurtu ‘s test pipeline should allow you to set something like that up. If you can figure out a better mechanism to do “load the new logs,” please do! I was never 100% satisfied with this even when it was working!

Reply:
I believe this is the bug that needs to be validated: #354.

Reply:
Hi @tscurtu 

Thanks for the reminder. This PR should fix #354 and I validated that.